sources/kms/grant.go

```diff
@@ -0,0 +1,220 @@
+		}
+
+		if grant.KeyId != nil {
+			arn, errA := sources.ParseARN(*grant.KeyId)
+			if errA != nil {
+				return nil, &sdp.QueryError{
+					ErrorType:   sdp.QueryError_OTHER,
+					ErrorString: fmt.Sprintf("failed to parse keyID: %s", *grant.KeyId),
+				}
+			}
+
+			// +overmind:link kms-key
+			item.LinkedItemQueries = append(item.LinkedItemQueries, &sdp.LinkedItemQuery{
+				Query: &sdp.Query{
+					Type:   "kms-key",
+					Method: sdp.QueryMethod_GET,
+					Query:  strings.TrimPrefix(arn.Resource, "key/"),
+					Scope:  scope,
+				},
+				BlastPropagation: &sdp.BlastPropagation{
+					// These are tightly linked
+					// Adding or revoking/retiring a grant can allow or deny permission to the KMS key for the grantee.
+					In:  true,
+					Out: true,
+				},
+			})
+		}
+
+		var principals []string
+		if grant.GranteePrincipal != nil {
+			principals = append(principals, *grant.GranteePrincipal)
+		}
+
+		if grant.RetiringPrincipal != nil {
+			principals = append(principals, *grant.RetiringPrincipal)
+		}
+
+		// Valid principals include
+		// - Amazon Web Services accounts
+		// - IAM users,
+		// - IAM roles,
+		// - federated users,
+		// - assumed role users.
+		// principals: https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_identifiers.html#identifiers-arns
+		/*
+			arn:aws:iam::account:root
+			arn:aws:iam::account:user/user-name-with-path
+			arn:aws:iam::account:role/role-name-with-path
+			arn:aws:sts::account:federated-user/user-name
+			arn:aws:sts::account:assumed-role/role-name/role-session-name
+			arn:aws:sts::account:self
+
+			dynamodb.us-west-2.amazonaws.com
+
+			The following are not supported
+				- arn:aws:iam::account:root
+				- arn:aws:sts::account:federated-user/user-name
+				- arn:aws:sts::account:assumed-role/role-name/role-session-name
+				- arn:aws:sts::account:self
+				- dynamodb.us-west-2.amazonaws.com => this will cause an error in ARN parsing
+		*/
+
+		for _, principal := range principals {
+			lIQ := &sdp.LinkedItemQuery{
+				Query: &sdp.Query{
+					Method: sdp.QueryMethod_GET,
+					Scope:  scope,
+				},
+				BlastPropagation: &sdp.BlastPropagation{
+					// These are tightly linked
+					// Adding or revoking/retiring a grant can allow or deny permission to the KMS key for the grantee.
+					// Or, disabling a role will make the grant redundant.
+					In:  true,
+					Out: true,
+				},
+			}
+
+			arn, errA := sources.ParseARN(principal)
+			if errA != nil {
+				return nil, &sdp.QueryError{
+					ErrorType:   sdp.QueryError_OTHER,
+					ErrorString: fmt.Sprintf("failed to parse principal: %s", principal),
+				}
+			}
+
+			switch arn.Service {
+			case "iam":
+				source, query := iamSourceAndQuery(arn.Resource)
+				switch source {
+				case "user":
+					// +overmind:link iam-user
+					lIQ.Query.Type = "iam-user"
+					lIQ.Query.Query = query
+				case "role":
+					// +overmind:link iam-role
+					lIQ.Query.Type = "iam-role"
+					lIQ.Query.Query = query
+				default:
```

Comment:
Maybe we can just ignore/log the error cases until we figure out what to do with the unsupported principals.